clients/da-vinci-client/src/main/java/com/linkedin/davinci/store/view/MaterializedViewWriter.java

```diff
@@ -0,0 +1,232 @@
+  private static final RedundantExceptionFilter REDUNDANT_LOGGING_FILTER =
+      RedundantExceptionFilter.getRedundantExceptionFilter();
+
+  public MaterializedViewWriter(
+      VeniceConfigLoader props,
+      Version version,
+      Schema keySchema,
+      Map<String, String> extraViewParameters,
+      Clock clock) {
+    super(props, version, keySchema, extraViewParameters);
+    pubSubProducerAdapterFactory = props.getVeniceServerConfig().getPubSubClientsFactory().getProducerAdapterFactory();
+    internalView =
+        new MaterializedView(props.getCombinedProperties().toProperties(), version.getStoreName(), extraViewParameters);
+    materializedViewTopicName =
+        internalView.getTopicNamesAndConfigsForVersion(version.getNumber()).keySet().stream().findAny().get();
+    this.clock = clock;
+  }
+
+  public MaterializedViewWriter(
+      VeniceConfigLoader props,
+      Version version,
+      Schema keySchema,
+      Map<String, String> extraViewParameters) {
+    this(props, version, keySchema, extraViewParameters, Clock.systemUTC());
+  }
+
+  /**
+   * package private for testing purpose
+   */
+  void setVeniceWriter(VeniceWriter veniceWriter) {
+    this.veniceWriter = veniceWriter;
+  }
+
+  @Override
+  public CompletableFuture<PubSubProduceResult> processRecord(
+      ByteBuffer newValue,
+      ByteBuffer oldValue,
+      byte[] key,
+      int newValueSchemaId,
+      int oldValueSchemaId,
+      GenericRecord replicationMetadataRecord) {
+    return processRecord(newValue, key, newValueSchemaId);
+  }
+
+  @Override
+  public CompletableFuture<PubSubProduceResult> processRecord(ByteBuffer newValue, byte[] key, int newValueSchemaId) {
+    if (veniceWriter == null) {
+      initializeVeniceWriter();
+    }
+    return veniceWriter.put(key, newValue.array(), newValueSchemaId);
+  }
+
+  @Override
+  public void processControlMessage(
+      KafkaKey kafkaKey,
+      KafkaMessageEnvelope kafkaMessageEnvelope,
+      ControlMessage controlMessage,
+      int partition,
+      PartitionConsumptionState partitionConsumptionState) {
+    final ControlMessageType type = ControlMessageType.valueOf(controlMessage);
+    // Ignore other control messages for materialized view.
+    if (type == ControlMessageType.START_OF_SEGMENT && Arrays.equals(kafkaKey.getKey(), KafkaKey.HEART_BEAT.getKey())) {
+      maybePropagateHeartbeatLowWatermarkToViewTopic(
+          partition,
+          partitionConsumptionState,
+          kafkaMessageEnvelope.getProducerMetadata().getMessageTimestamp());
+    }
+  }
+
+  @Override
+  public String getWriterClassName() {
+    return internalView.getWriterClassName();
+  }
+
+  // Package private for testing
+  VeniceWriterOptions buildWriterOptions() {
+    return setProducerOptimizations(internalView.getWriterOptionsBuilder(materializedViewTopicName, version)).build();
+  }
+
+  synchronized private void initializeVeniceWriter() {
```

Comment:
both ChangeCaptureViewWriter and this writer have the same initWriter method. Can we pull them to their common parent class? 

Also, this may be an ask that went too far: instead of using the heavy weight of synchronized, I wonder whether using an AtomicReferenceFieldUpdater with which you don't have to change viewWriter? Or using an atomicReference, with which you will have to change the viewWriter to Reference of ViewWriter? 

If the init method is not involved frequently, maybe it is not worthwhile to do such a refactor.

Reply:
The concurrency aspect of init is only called for the very first time since we do null check before calling init in subsequent calls. I should be able to refactor the method to their parent class `VeniceViewWriter`

Reply:
Thanks! I understand that those double nullity check is ugly but it is an industrial norm. It is totally fine for us to live with the "synchronized".